lib/linkerd2-error/src/recover.rs

```diff
@@ -0,0 +1,63 @@
+use super::{Error, Never};
+use futures::{stream, Stream};
+
+/// An error recovery strategy.
+pub trait Recover<E: Into<Error> = Error> {
+    type Error: Into<Error>;
+    type Backoff: Stream<Item = (), Error = Self::Error>;
+
+    /// Given an E-typed error, determine if the error is recoverable.
+    ///
+    /// If it is, a backoff stream is returned. When the backoff becomes ready,
+    /// it signals that the caller should retry its operation. If the backoff is
+    /// polled agian, it is assumed that the operation failed and a new (possibly
+    /// longer) backoff is initated.
+    ///
+    /// If the error is not recoverable, it is returned immediately.
+    fn recover(&self, err: E) -> Result<Self::Backoff, E>;
```

Comment:
I understand your point but I don't know that I'm especially bothered by the current API... because backoff is lazy, it's basically free to return an unused backoff. I also like that this API allows the implementor to modify the Error type.

I'm -1 on `next_backoff` as proposed there, as there's nothing that requires a backoff strategy be strictly delay-based, and stream hides that detail nicely. Furthermore this would expose a much more stateful API, as the caller would have to inform the recover when the operation has been completed successfully! My proposed API does this all with RAII.

Reply:
I'm much less concerned about efficiency or allocations and much more worried about this being an API that is easy to use incorrectly.  A naive user (like myself) could easily call `recover` and use the result for each error, thus never actually backing off.  Or, I could just as easily assume that the backoff that I have is what I should use, but forget that I still need to call `recover` on each error to check that the error is recoverable.

Reply:
> A naive user (like myself) could easily call recover and use the result for each error, thus never actually backing off.

Why would you think it appropriate to drop the first backoff stream in favor of the new one? That seems like an arbitrary, though intentional, decision. If we're saying that the docs can be improved to make the contract clearer, I agree. But as far as I can tell the other suggestions would have the same problem.

When recover gives you a backoff, you have to think "when should I poll this" and "when should I drop this". The answers to these questions are not ambiguous, though they may need to be more clearly spelled out in the docs.

My main argument is that if we changed this to be a more imperative style API where we had to call `set_recovered()` or some such to indicate that the backoff had completed, it would be just as easy to accidentally do the wrong thing if you don't read the full API contract. By virtue of returning a Stream, we can use the type system (as in reconnect) to be more clear about how the contract is enforced (for instance, it's impossible to reuse a backoff after a service has been connected successfully).

To eliza's point, we could try to decouple "checking" from creating the backoff, but it would proably need to be a bit funk, as well, in order to preserve some of the properties of the current api:

```rust
trait Recover<E> {
    type Error: Into<Error>;
    type Backoff: Stream<Item = (), Error = Self::Error>;

    fn check_recoverable(&self, err: E) -> Result<E, E>;
    fn new_backoff(&self, err: E) -> Self::Backoff;
}
```

That seems fine, I guess? but is it substantially better?

Reply:
I'm fine with this API, I just wanted to give my perspective.

That said, I don't think I've clearly communicated my suggestion.  My intention is that `recover` would return a backoff object that has a `next_backoff` method.  The backoff object would be stateful, just like Stream is.  This utilizes RAII just like the current implementation does.  The difference is that `recover` is only called for the first error (i.e. when creating a new backoff) and then `next_backoff` is called on the backoff object for each error.  When service creation succeeds, the backoff object is dropped and the next error can call `recover` to create a new one.

To preserve the hiding of the backoff implementation, we could use a slightly different trait from Stream:

```
trait BackoffStream { 
  fn poll_backoff(&mut self, err: E) -> Result<Async<()>, E>;
}
```

But this has the problem that you have to clone the E each time you poll.  So really you'd want something more like...

```
trait BackoffStream {
  type Future: Future<Item=()>;
  fn next_backoff(&mut self, err: E) -> Result<Self::Future, E>;
}
```

But anyway, just sharing my thought process.

Reply:
@adleong yeah, i totally see where you're going. FWIW, I considered something like this but what we have here felt like the right balance. I reserve the right to change my mind at any point in the future ;p